src/transform-sdk/rust/src/lib.rs

```diff
@@ -0,0 +1,166 @@
+///
+/// use redpanda::*;
+///
+/// fn main() {
+///   on_record_written(my_transform);
+/// }
+///
+/// fn my_transform(event: WriteEvent) -> Result<Vec<Record>> {
+///   Ok(vec![
+///     Record::new(
+///       event.record.key().map(|k| k.to_owned()),
+///       event.record.value().map(|v| v.to_owned()),
+///     )
+///   ])
+/// }
+/// ```
+pub fn on_record_written(cb: OnRecordWritten) {
+    process(cb)
+}
+
+fn process(cb: OnRecordWritten) {
+    unsafe {
+        abi::check_abi();
+    }
+    loop {
+        process_batch(cb);
+    }
+}
+
+fn process_batch(cb: OnRecordWritten) {
+    let mut header = record::BatchHeader {
+        base_offset: 0,
+        record_count: 0,
+        partition_leader_epoch: 0,
+        attributes: 0,
+        last_offset_delta: 0,
+        base_timestamp: 0,
+        max_timestamp: 0,
+        producer_id: 0,
+        producer_epoch: 0,
+        base_sequence: 0,
+    };
+    let errno_or_buf_size = unsafe {
+        abi::read_batch_header(
+            &mut header.base_offset,
+            &mut header.record_count,
+            &mut header.partition_leader_epoch,
+            &mut header.attributes,
+            &mut header.last_offset_delta,
+            &mut header.base_timestamp,
+            &mut header.max_timestamp,
+            &mut header.producer_id,
+            &mut header.producer_epoch,
+            &mut header.base_sequence,
+        )
+    };
+    assert!(
+        errno_or_buf_size >= 0,
+        "failed to read batch header (errno: {errno_or_buf_size})"
+    );
+    let buf_size = errno_or_buf_size as usize;
+    let mut input_buffer: Vec<u8> = vec![0; buf_size];
+    let mut output_buffer: Vec<u8> = Vec::new();
+    for _ in 0..header.record_count {
+        let mut attr: u8 = 0;
+        let mut timestamp: i64 = 0;
+        let mut offset: i64 = 0;
+        let errno_or_amt = unsafe {
+            abi::read_next_record(
+                &mut attr,
+                &mut timestamp,
+                &mut offset,
+                input_buffer.as_mut_ptr(),
+                input_buffer.len() as u32,
+            )
+        };
+        assert!(
+            errno_or_amt >= 0,
+            "reading record failed (errno: {errno_or_amt}, buffer_size: {buf_size})"
+        );
+        let amt = errno_or_amt as usize;
+        let ts = SystemTime::UNIX_EPOCH + Duration::from_millis(timestamp as u64);
+        let record = BorrowedRecord::read_from_payload(&input_buffer[0..amt], ts)
+            .expect("deserializing record failed");
+        let transformed = cb(WriteEvent { record }).expect("transforming record failed");
+        for record in transformed {
+            output_buffer.clear();
+            record.write_payload(&mut output_buffer);
+            let errno_or_amt =
+                unsafe { abi::write_record(output_buffer.as_ptr(), output_buffer.len() as u32) };
+            assert!(
+                errno_or_amt as usize == output_buffer.len(),
```

Comment:
is output_buffer guaranteed to be larger than any possibly defined errno? :)

Reply:
wasm returns an `i32` and so you'd have to have a bigger size here than `i32::MAX`, which will certainly OOM (wasm has a 32 bit address space)

Reply:
@rockwotj i read `pub(crate) fn write_record(buf: *const u8, len: u32) -> **u**32;` typo?

Reply:
> @rockwotj i read `pub(crate) fn write_record(buf: *const u8, len: u32) -> **u**32;` typo?

Ah yes thanks it is a typo! Fixed.